### DISCUSSION TITLE
helix on Chrome ?

### DISCUSSION THREAD

**Discussion by ide5kllc** [helix-editor/helix#4706] [Q&A]
I am trying to build a fully in-browser IDE.

I am looking for an implemented-in-Rust editor with LSP support that can build on wasm32-unknown-unknown.

Is there currently anyway to build Helix for wasm32-unknown-unknown and run it in Chrome ?

It is okay if there is no file system access and I can only edit an in-memory buffer.

**Comment by the-mikedavis** [comment]
@archseer put some work into getting compilation working for WASM (so maybe he can say more here) but AFAIK Helix is not currently usable when compiled as WASM

**Reply by ide5kllc** [reply]
Is this "compiles on wasm32" code publicly accessible?

I have code that renders text via a WebGL2 fragment shader implementation of signed-distance-fields.

Thus, if Helix compiles on wasm32, I may be able to get something useful out of it.

**Reply by kirawi** [reply]
It's in master.

**Reply by ide5kllc** [reply]
git clone https://github.com/helix-editor/helix
cargo build --target=wasm32-unknown-unknown

is giving me all types of error msgs regarding crates: bytes, unicode-bidi, ...

Is there a branch / tag / ... something I need to do to build for wasm32 ?

**Reply by kirawi** [reply]
It's probably not far along enough to use on the web.

**Reply by ide5kllc** [reply]
I have a WebGL2 fragment shader that renders text via SDF.
I also have a LSP server for https://github.com/gluon-lang/gluon running in a webworker on Chrome/wasm32.
I am looking to replace CodeMirror dependency.

If I can just get Helix to BUILD for wasm32-unknown-unknown, I can probably monkey patch the rest to get it to render on screen.

**Reply by ide5kllc** [reply]
Where are the instructions / branch / tag for building helix on wasm32 ?

**Reply by the-mikedavis** [reply]
Ah, I was thinking of https://github.com/helix-editor/helix/commit/9191af3f8d73d2101c46192a4a80c41a19012320. So some of the helix crates could compile for wasm (note: that commit is from a while ago so things may have changed) but overall Helix does not currently compile for wasm.

**Reply by archseer** [reply]
Yeah it's not easy to do this right now without a lot of work and a separate frontend built in webGL. Alternatively `helix-term` frontend could probably run on `xterm.js` but that's also a lot of work.

**Reply by archseer** [reply]
There's a severely outdated attempt that ports more of the code but disables LSP https://github.com/helix-editor/helix/tree/gui

**Reply by ide5kllc** [reply]
Oh, I want LSP. I invested time learning about LSP internals and getting https://github.com/gluon-lang/gluon_language-server running in a webworker and talking to brower over postMessage. I'm not giving up LSP.

I am interested in hearing about "helix-term" on "xterm.js" -- I stumbled across xterm.js, but never gave it more thought. Is it conceivable that there is some crate TERM_EMULATOR that helix-term depends on, which currently does not compile on wasm32, but which we can fork, and stub in xterm.js in it?

Then, we take whatever API that TERM_EMULATOR exports, and using some Rust "ffi" (to JS) calls route it to xterm.js ?

**Reply by ide5kllc** [reply]
Not tested: https://github.com/segeljakt/xterm-js-rs claims to have Rust bindings to xterm-js.

Is this as "simple" as routing TERM_EMULATOR api to xterm-js-rs, or am I missing something important ?

**Reply by ide5kllc** [reply]
Another untested library: https://docs.rs/crate/xterm-js-sys/4.5.0-alpha0

**Comment by makemeunsee** [comment]
Hello,

I got curious and tried to 'resurrect' the `gui` branch and compile (parts of) Helix to `wasm32`.
That was a learning experience but it was too messy.
Even trying from promising commits like [helix-view now builds on WASM](https://github.com/helix-editor/helix/commit/6f8bf67fa6c3bf8d1d9995b9bda46a3197664803) or [it compiles!](https://github.com/helix-editor/helix/commit/d1515671928d577481d7a77c5fef25b1f1b157fb).

Anyway, as I am only barely discovering the project, I have some questions about this whole ordeal:

* what component/aspect of `lsp`  makes it (currently?) incompatible with the `wasm32` target?
* what component/aspect of debugging (`dap`) makes it (currently?) incompatible with the `wasm32` target?
* in the current state of Helix' code (as opposed to 2 years ago when the `gui` branch was created), what would be the high-level steps to port the code to `wasm32`? (e.g. decoupling `helix-view` from the terminal UI)

I get that `tokio` is a big offender in this regard, at least some features of it (see [`wasm` supported features](https://docs.rs/tokio/latest/tokio/#wasm-support)), so I'm interested in learning about the other ones.

Also for `tree-sitter`, the `tree-sitter-facade` and `web-tree-sitter-sys` that _archseer_ was mentioning in the `gui` branch are now archived, but [the Rust port of `tree-sitter`](https://github.com/shadaj/tree-sitter-c2rust) could be a better replacement (see [this discussion](https://github.com/tree-sitter/tree-sitter/discussions/1550)).

Thanks for any input on this topic!

**Reply by makemeunsee** [reply]
I dove a bit into this and made decent progress, enough to warrant a self-reply I think:

* language servers and debugging tools are fully native, external processes, so are incompatible with `wasm32` by nature, and integrating with hypothetical `wasm32` compatible tools would require a significant refactoring of `lsp` and `dap` designs within Helix.
* massive feature gating is enough to enable compilation to `wasm32`; the challenge is to design and implement the integration with a web terminal backend.
* [`tree-sitter-c2rust`](https://github.com/shadaj/tree-sitter-c2rust) works almost as a drop in replacement for `tree-sitter`, which is nice. How to provide the grammars remains to be solved.

Good news is I got the whole project (minus excluded features) to build to `wasm32`, but there remains a lot to be done to make it usable.

This is the branch where it lives: https://github.com/makemeunsee/helix/tree/wasm32
